src/poll.rs

```diff
@@ -52,14 +90,21 @@ const MAX_REFCOUNT: usize = (isize::MAX) as usize;
 /// poll.poll(&mut events, None).unwrap();
 /// ```
 pub struct Poll {
-    // This type is `Send`, but not `Sync`, so ensure it's exposed as such.
-    _marker: marker::PhantomData<Cell<()>>,
-
     // Platform specific IO selector
     selector: sys::Selector,
 
     // Custom readiness queue
     readiness_queue: ReadinessQueue,
+
+    // Use an atomic to first check if a full lock will be required. This is a
+    // fast-path check for single threaded cases avoiding the extra syscall
+    lock_state: AtomicUsize,
```

Comment:
I find this fast-path check dubious. Does it really have any significant performance benefits? 

I’m asking this, firstly because the `Mutex` is supposed to do this itself – user-space mutexes are supported on all major systems except macOS (which still does not syscall if no contention is present anyway) and `Mutex` impl should already be using them. And secondly, because implementing locks correctly is hard and if we absolutely must do this, its probably better to just use [parking_lot](https://crates.io/crates/parking_lot).